analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/NullPointerDereferenceBase.cs

```diff
@@ -20,34 +20,45 @@
 {
-    public abstract class NullPointerDereferenceBase : SymbolicRuleCheck
-    {
-        internal const string DiagnosticId = "S2259";
+    internal const string DiagnosticId = "S2259";
 
-        protected virtual bool IsSupressed(SyntaxNode node) => false;
+    protected virtual bool IsSupressed(SyntaxNode node) => false;
 
-        protected override ProgramState PreProcessSimple(SymbolicContext context)
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        if (ReferenceOrDefault(context.Operation.Instance) is { Syntax: var syntax} reference
+            && context.HasConstraint(reference, ObjectConstraint.Null)
+            && !IsSupressed(syntax)
+            && SemanticModel.GetTypeInfo(syntax).Nullability().FlowState != NullableFlowState.NotNull)
         {
-            var reference = context.Operation.Instance.Kind switch
-            {
-                OperationKindEx.Invocation => context.Operation.Instance.ToInvocation().Instance,
-                OperationKindEx.PropertyReference => context.Operation.Instance.ToPropertyReference().Instance,
-                OperationKindEx.Await => context.Operation.Instance.ToAwait().Operation,
-                OperationKindEx.ArrayElementReference => context.Operation.Instance.ToArrayElementReference().ArrayReference,
-                _ => null,
-            };
-            if (reference != null
-                && context.HasConstraint(reference, ObjectConstraint.Null)
-                && !reference.Type.IsStruct() // ToDo: IsStruct() is a workaround before MMF-2401
-                && !IsSupressed(reference.Syntax)
-                && SemanticModel.GetTypeInfo(reference.Syntax).Nullability().FlowState != NullableFlowState.NotNull)
-            {
-                ReportIssue(reference, reference.Syntax.ToString());
-            }
-
-            return context.State;
+            ReportIssue(reference, syntax.ToString());
         }
+        return context.State;
     }
+
+    private static IOperation ReferenceOrDefault(IOperation operation) =>
+        operation.Kind switch
+        {
+            OperationKindEx.Invocation => InvocationOrDefault(operation),
+            OperationKindEx.PropertyReference => PropertyReferenceOrDefault(operation),
+            OperationKindEx.Await => operation.ToAwait().Operation,
+            OperationKindEx.ArrayElementReference => operation.ToArrayElementReference().ArrayReference,
+            _ => null,
+        };
+
+    private static IOperation InvocationOrDefault(IOperation operation) =>
+        operation.ToInvocation() is { Instance: var instance, TargetMethod: var method }
+        && (!method.ContainingType.Is(KnownType.System_Nullable_T) || method.Name == nameof(Nullable<int>.GetType)) // All methods on Nullable but .GetType() are safe to call
+            ? instance
+            : null;
+
+    private static IOperation PropertyReferenceOrDefault(IOperation operation) =>
```

Comment:
The combination of preconditions (assumptions), build-in conditions and name doesn't sound right.

Here, we're 100% sure we're working with property reference => it should have `IPropertyReferenceOperationWrapper property` as an argument. And the invocation side should do the `ToPropertyReference()` because it also ensures the condition via switch expression.

Then we won't need the null check, nor pattern matching. 

Then we'll be left just with the `Is nullable` check. And we should adjust the name accordingly. Something like `NonNullablePropertyReference`.